tools/autograd/templates/Functions.cpp

```diff
@@ -2321,6 +2321,129 @@ std::tuple<Tensor, Tensor, Tensor> batchnorm_double_backward(
 
 }
 
+std::tuple<Tensor, Tensor, Tensor>
+infinitely_differentiable_native_layer_norm_backward(
+    Tensor grad_out,
+    const Tensor& grad_mean,
+    const Tensor& grad_rstd,
+    const Tensor& input,
+    const Tensor& mean,
+    const Tensor& rstd,
+    const Tensor& weight,
+    int64_t M,
+    int64_t N,
+    double eps,
+    std::array<bool, 3> grad_input_mask) {
+  Tensor grad_input;
+  Tensor grad_weight;
+  Tensor grad_bias;
+
+  Tensor flattened_input = input.reshape({M, N});
+  Tensor flattened_weight;
+  if (weight.defined()) {
+    flattened_weight = weight.reshape({N});
+  }
+
+  // Common subexpressions
+  Tensor input_minus_mean_times_grad_out;
+  if (grad_out.defined()) {
+    grad_out = grad_out.reshape({M, N});
+    input_minus_mean_times_grad_out =
+        (flattened_input - mean.unsqueeze(1)) * grad_out;
+  }
+
+  if (grad_input_mask[0]) {
+    // Let's say
+    // out, mean, rstd = layernorm(input, weight, bias) = (input - mean) * rstd
+    // * weight + bias := f(input, mean, rstd, weight, bias) y = g(out, mean,
```

Comment:
nit: Should we put `y = ...` in a new line to be more clear?

Reply:
> Should we benchmark on `(M, N)` where `M` could be as small as `8` and as large as `4096`, and `N` being `768`, `1024`, `1280`, `2048`? These are the real-world use case I see in https://github.com/huggingface/transformers

The reason I didn't test input with such small size is the overhead in the ops will become comparable large while the result is not very stable.

One of the test with such small input is here https://gist.github.com/BIT-silence/4c762eb6c4db55d58a57a43f492674df
While the timeit result cannot match the profiler result here https://gist.github.com/BIT-silence/27848840dc77f26fc459c8cc5311eb65.
And from the profiler result, we can see that things like empty_like will take a large percentage of time.

Here is the profiler result for the large input https://gist.github.com/BIT-silence/ba01562912bf5815a8059093a87f7c15.
We can see for the large inputs, the numbers from timeit and profiler can matches.

BTW, I found I made a mistake when doing the benchmark that I included the gc.collect time for the 1000 runs, I have changed the benchmark script and update the results above.

Reply:
Formatted.